### PR TITLE
genericapiserver: write negotiated Status on timeout

### PR DESCRIPTION
**Pull request by sttts**
~~**Blocked** by pkg/genericapiserver/endpoints not moved to k8s.io/apiserver yet.~~

This PR 
- simplifies the timeout filter by using less indirections, 
- adds negotiated `meta.Status` errors on actual timeout,
- adds tests with real `RequestInfo` and `RequestContextMapper`.

### CODE REVIEW

File: pkg/genericapiserver/server/filters/timeout.go [hunk 1]
```diff
@@ -18,92 +18,98 @@ package filters
...
+				responsewriters.InternalError(w, r, errors.New("no context found for request"))
+			}
+
+			requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+			if !ok {
+				responsewriters.InternalError(w, r, errors.New("no request info found for request"))
+			}
+
+			gvr := schema.GroupVersionResource{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion, Resource: requestInfo.Resource}
+			if gvr.Version == "" {
```
Comment (sttts) [inline pkg/genericapiserver/server/filters/timeout.go]: @smarterclayton is this correct? We return a `v1` Status if there is no version in the `RequestInfo`. This probably means that the URL path is outside of a versioned group path.

Comment (smarterclayton) [inline pkg/genericapiserver/server/filters/timeout.go]: Yes, v1.Status is the default for all older API groups and for anything outside of the groups.

File: pkg/genericapiserver/server/filters/timeout.go [hunk 2]
```diff
@@ -18,92 +18,98 @@ package filters
...
 	}
-	timeoutFunc := func(req *http.Request) (<-chan time.Time, string) {
-		// TODO unify this with apiserver.MaxInFlightLimit
+	handlerWithTimeout := WithTimeout(handler, globalTimeout, requestContextMapper, s)
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 		ctx, ok := requestContextMapper.Get(req)
 		if !ok {
-			return time.After(globalTimeout), ""
+			responsewriters.InternalError(w, req, errors.New("no context found for request"))
 		}
```
Comment (sttts) [inline pkg/genericapiserver/server/filters/timeout.go]: return

File: pkg/genericapiserver/server/filters/timeout.go [hunk 3]
```diff
@@ -18,92 +18,98 @@ package filters
...
-	}
+// handler responds with a 504 StatusGatewayTimeout error.
+func WithTimeout(handler http.Handler, timeout time.Duration, requestContextMapper apirequest.RequestContextMapper, s runtime.NegotiatedSerializer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		timedout := time.After(timeout)
+		done := make(chan struct{})
+		tw := newTimeoutWriter(w, func() {
+			ctx, ok := requestContextMapper.Get(r)
+			if !ok {
+				responsewriters.InternalError(w, r, errors.New("no context found for request"))
```
Comment (sttts) [inline pkg/genericapiserver/server/filters/timeout.go]: return

Comment (smarterclayton) [inline pkg/genericapiserver/server/filters/timeout.go]: Are you reviewing your own code?  Impressive!

**Comment by ethernetdan** [comment]
Removing issue from 1.6 milestone since code freeze has begun:
* If this is a bug fix, please label as `kind/bug` and reapply the milestone.
* If this is a feature for 1.6, an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) should be filed for it before 11:00PM PST on 3/2. Note that most exceptions to code freeze are not accepted.
* If this is a feature for 1.7, feel free to add the 1.7 milestone to the PR.

/cc @kubernetes/release-team @kubernetes/kubernetes-release-managers

**Comment by calebamiles** [comment]
Moving this PR into the 1.7 milestone @sttts  

**cc: @ethernetdan, @kubernetes/kubernetes-release-managers**

**Comment by smarterclayton** [comment]
Oops, completely forgot about this

**Comment by sttts** [comment]
> Oops, completely forgot about this

probably worth a rebase after freeze.

**Comment by cblecker** [comment]
@sttts @smarterclayton Is this PR still active? There doesn't seem to be meaningful action in over a year.
I'm going to remove the lifecycle/frozen status and allow this to close in ~120 days if there is no further action.

/remove-lifecycle frozen

**Comment by sttts** [comment]
@cblecker it is still broken and should be fixed.

**Pull request closed without merging**